### PR TITLE
Implement a meta-data store

### PR DESCRIPTION
**Pull request by rgraff**
This is a work in progress, see the readme changes for my intent

### CODE REVIEW

File: lib/event_sourceable.rb [hunk 1]
```diff
@@ -3,4 +3,15 @@
 require "event_sourceable/event_store"
 
 module EventSourceable
+  @metadata = {}
```
Comment (rgraff) [inline lib/event_sourceable.rb]: Based on discussion with @mpeychich this is not thread safe and probably won't work as implemented.

Consider using thread local store which I'm not familiar with.  Found this on the topic: https://www.rubytapas.com/2016/11/20/ruby-thread-local-variables/

**Review by rgraff — COMMENTED** [review]
Nice.

@mpeychich how's this look?

File: README.md [hunk 2]
```diff
@@ -196,18 +196,83 @@ end
...
+  def reactors
+    "User::SendConfirmationEmail"
+  end
+
+  def async_reactors
+    [User::SendConfirmationEmail, User::SendConfirmationEmail]
+  end
+end
+```
+
```
Comment (rgraff) [inline README.md]: I wrote this readme but never implemented it. Probably not how we'll actually implement.

```suggestion
```

File: lib/event_sourceable/event_store.rb [hunk 3]
```diff
@@ -13,7 +15,7 @@ class EventStore < ::ActiveRecord::Base
 
     after_initialize do
       self.data ||= {}
-      self.metadata ||= {}
+      self.metadata ||= EventSourceable.metadata
```
Comment (kroehre) [inline lib/event_sourceable/event_store.rb]: Should this be `dup`'d?

File: lib/event_sourceable/event_store.rb [hunk 4]
```diff
@@ -73,7 +75,7 @@ def apply_event_and_save
 
       # record the metadata
       if event.respond_to?(:metadata)
-        self.metadata.merge!(event.metadata || {})
+        self.metadata = metadata.merge!(event.metadata || {})
```
Comment (kroehre) [inline lib/event_sourceable/event_store.rb]: Should this be `merge`? `merge!` changes in place.

Comment (mpeychich) [reply lib/event_sourceable/event_store.rb]: I'd `merge` here without the bang

File: lib/event_sourceable.rb [hunk 5]
```diff
@@ -3,4 +3,17 @@
 require "event_sourceable/event_store"
 
 module EventSourceable
-end
\ No newline at end of file
+  def self.metadata
+    Thread.current[:event_sourceable_metadata] ||= {}
```
Comment (mpeychich) [inline lib/event_sourceable.rb:7]: `Thread.current` _may_ not give you the level of protection you want here. You will be fine in instances where the thread is not reused between contexts however, the thread is reused between contexts in environments like Puma or Sidekiq. The `around_action` ensures the resetting between requests for web but you will need something equivalent for sidekiq.  We can build this out more or just use something like:

https://github.com/steveklabnik/request_store
https://github.com/madebylotus/request_store-sidekiq

Comment (kroehre) [reply lib/event_sourceable.rb:7]: IIRC request store is essentially the Thread.current[] concept we have here but with auto-reset, yeah?

Comment (mpeychich) [reply lib/event_sourceable.rb:7]: yes, it is thread current with a middleware that resets it between requests. The sidekiq gem adds a server middleware that resets between jobs.

File: lib/event_sourceable/event_store.rb (hunk 3 above)
Comment (rgraff) [reply lib/event_sourceable/event_store.rb]: yes. let's dupe it just to prevent unintended changes by the event.

**Pull request closed without merging**